### PR TITLE
Support different extend modes using "extend.extend" config item

### PR DESCRIPTION
**Pull request by janaz**
This is my attempt at implementing the feature asked for in #3534  

@lovell suggested the following API for it:
```js
// PROPOSED API, NOT YET AVAILABLE
sharp.extend({
  ...
  extend: 'copy' | 'mirror' | 'repeat'
})
// PROPOSED API, NOT YET AVAILABLE
```

This PR adds a new config item to `sharp.extend` as specified in the proposal. The supported values for `extend.extend` are:

* `background` (colour set from the _background_ property) - this is the current behaviour (default)
extend-equal-background

* `copy` (copy the image edges)
extend-equal-copy

* `repeat` (repeat the whole image)
extend-equal-repeat

* `mirror` (mirror the whole image)
extend-equal-mirror

### CODE REVIEW

**Comment by lovell** [comment]
Hi Tomasz, good to hear from you, thank you very much for this.

It's interesting that you chose `ExtendMode` as the type name, perhaps we should

File: lib/index.d.ts [hunk 1]
```diff
@@ -1243,6 +1243,8 @@ declare namespace sharp {
         sigma?: number | undefined;
     }
 
+    type ExtendMode = 'background' | 'copy' | 'repeat' | 'mirror';
```
Comment (lovell) [inline lib/index.d.ts:1246]: Sorry for the API naming confusion, which is my fault as I suggested it in the first place, but I think this type would be less confusing when named `ExtendWith`.

**Review by lovell — CHANGES_REQUESTED** [review]
This is brilliant, thank you again Tomasz. I've left a couple of comments inline and I've given the public API some more thought too with one small change, but otherwise this looks great and is almost ready to go.

File: src/pipeline.cc [hunk 2]
```diff
@@ -1499,6 +1518,7 @@ Napi::Value pipeline(const Napi::CallbackInfo& info) {
   baton->extendLeft = sharp::AttrAsInt32(options, "extendLeft");
   baton->extendRight = sharp::AttrAsInt32(options, "extendRight");
   baton->extendBackground = sharp::AttrAsVectorOfDouble(options, "extendBackground");
+  baton->extendMode = sharp::AttrAsInt32(options, "extendMode");
```
[the remark this answers is not in the document]
Comment (janaz) [reply src/pipeline.cc:1521]: Thanks @lovell. I agree that using an enum type is better in this case.

File: lib/resize.js [hunk 3]
```diff
@@ -393,6 +405,13 @@ function extend (extend) {
       }
     }
     this._setBackgroundColourOption('extendBackground', extend.background);
+    if (is.defined(extend.extend)) {
```
[the remark this answers is not in the document]
Comment (janaz) [reply lib/resize.js:408]: 👍

File: lib/index.d.ts (hunk 1 above)
Comment (janaz) [reply lib/index.d.ts:1246]: No worries. It was clearly stated that it was just a proposal. Names of public API attributes are hard to change and now is the best time to think about it.

**Comment by janaz** [comment]
I pushed a commit with all the requested changes. Switching to `sharp::AttrAsEnum<VipsExtend>` made the code a lot simpler!

File: src/operations.cc [hunk 4]
```diff
@@ -395,11 +395,11 @@ namespace sharp {
    * Split into frames, embed each frame, reassemble, and update pageHeight.
    */
   VImage EmbedMultiPage(VImage image, int left, int top, int width, int height,
-                        std::vector<double> background, int nPages, int *pageHeight) {
+                        VipsExtend extendWith, std::vector<double> background, int nPages, int *pageHeight) {
     if (top == 0 && height == *pageHeight) {
       // Fast path; no need to adjust the height of the multi-page image
       return image.embed(left, 0, width, image.height(), VImage::option()
-        ->set("extend", VIPS_EXTEND_BACKGROUND)
+        ->set("extendWith", extendWith)
```
Comment (lovell) [inline src/operations.cc:402]:
```suggestion
        ->set("extend", extendWith)
```

**Review by lovell — CHANGES_REQUESTED** [review]
One small error but otherwise this is good to go, thank you.

**Review by lovell — APPROVED** [review]

**Comment by lovell** [comment]
Thanks again, this will be in v0.32.0.

**Pull request merged**